File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/IngredientService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.IngredientGroupDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.IngredientListDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.IngredientSuggestionDto;
import at.ac.tuwien.sepr.groupphase.backend.exception.ConflictException;
import at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;

import java.util.List;

/**
 * Service for Ingredients Entity.
 */
public interface IngredientService {

    /**
     * Searching for ingredients.
     *
     * @param ingredientsName name of ingredients
     * @return ingredients with searched name
     * @throws JsonProcessingException in case of api exception
     * @throws JsonMappingException    in case of mapping exception
     */
    List<IngredientListDto> searchIngredients(String ingredientsName) throws JsonMappingException, JsonProcessingException;

    /**
     * Get all ingredients.
     *
     * @return all ingredients
     */
    List<IngredientGroupDto> getAllGroupIngredients(Long groupId) throws NotFoundException;

    /**
     * Retrieve all stored ingredients, that match the given parameters.
     * The parameters may include a limit on the amount of results to return.
     *
     * @param searchParams parameters to search ingredients by
     * @return a stream of ingredients matching the parameters
     */
    List<IngredientListDto> searchUserIngredients(String searchParams);

    /**
     * Retrieve all stored ingredients, are associated with a user.
     *
     * @return a stream of ingredients belonging to a user
     */
    List<IngredientListDto> getUserIngredients();

    /**
     * Add ingredients to user.
     *
     * @param ingredientListDto ingredients to add
     */
    List<IngredientListDto> addIngredientsToUser(IngredientListDto[] ingredientListDto) throws ConflictException;

    /**
     * Get ingredient suggestions for a group.
     *
     * @param groupId id of the group
     * @return list of ingredient suggestions
     * @throws NotFoundException if the group does not exist
     * @throws ConflictException if the current user is not a member of the group or not the host
     */
    List<IngredientSuggestionDto> getIngredientSuggestions(Long groupId) throws NotFoundException, ConflictException;
}

File: frontend/src/app/services/ingredient.service.ts
import {Injectable} from '@angular/core';
import {Globals} from '../global/globals';
import {HttpClient} from "@angular/common/http";
import {IngredientGroupDto, IngredientListDto, IngredientSuggestionDto} from "../dtos/ingredient";
import {Observable, switchMap} from "rxjs";

@Injectable({
    providedIn: 'root'
})
export class IngredientService {

    private ingredientBaseUri: string = this.globals.backendUri + '/ingredients';

    constructor(
        private httpClient: HttpClient,
        private globals: Globals
    ) {
    }

    /**
     * Loads all ingredients of a group from the backend
     */
    getAllGroupIngredients(groupId: number): Observable<IngredientGroupDto[]> {
        return this.httpClient.get<IngredientGroupDto[]>(this.ingredientBaseUri + `/${groupId}`);
    }

    /**
     * Returns ingredients suggestions for a group
     *
     * @param groupId The id of the group to get suggestions for
     */
    getIngredientSuggestions(groupId: number): Observable<IngredientSuggestionDto[]> {
        return this.httpClient.get<IngredientSuggestionDto[]>(this.ingredientBaseUri + `/suggestions/${groupId}`);
    }


    /**
     * Adds an ingredient to the current user's ingredients
     *
     * @param ingredient The ingredient to add
     */
    addIngredientSuggestion(ingredient: IngredientListDto): Observable<IngredientListDto[]> {
        return this.getUserIngredients().pipe(
            switchMap((ingredients: IngredientListDto[]) => {
                ingredients.push(ingredient);
                return this.saveUserIngredients(ingredients);
            })
        );
    }

    /**
     * Loads all fitting ingredients for autocomplete
     */
    public searchIngredientsUserExisting(name: string): Observable<IngredientListDto[]> {
        return this.httpClient.get<IngredientListDto[]>(this.ingredientBaseUri + '/user-ingredients-auto/' + name);
    }

    /**
     * Loads all ingredients of a user from the backend
     */
    getUserIngredients(): Observable<IngredientListDto[]> {
        return this.httpClient.get<IngredientListDto[]>(this.ingredientBaseUri + '/user-ingredients');
    }

    /**
     * Adds ingredients to the backend which the user had set
     */
    saveUserIngredients(ingredients: IngredientListDto[]): Observable<IngredientListDto[]> {
        return this.httpClient.post<IngredientListDto[]>(this.ingredientBaseUri + '/user-ingredients', ingredients);
    }

    /**
     * Searching for ingredients
     *
     * @param name of ingredient
     * @returns return matched ingredients
     */
    search(name: String): Observable<IngredientListDto[]> {
        if (name == "") {
          return this.httpClient.get<IngredientListDto[]>(`${this.ingredientBaseUri}/searchIngredients/` + "null");
        } else {
          return this.httpClient.get<IngredientListDto[]>(`${this.ingredientBaseUri}/searchIngredients/` + name);
        }
    }

}
